File: java/src/main/java/toolgood/algorithm/mathNet/Distributions/Binomial.java
package toolgood.algorithm.mathNet.Distributions;

import java.math.BigDecimal;
import java.math.MathContext;

import toolgood.algorithm.system.MathEx;
import toolgood.algorithm.mathNet.SpecialFunctions;

public class Binomial {
    public static BigDecimal PMF(BigDecimal p, int n, int k) {
        if (k < 0 || k > n) {
            return BigDecimal.ZERO;
        }

        if (p.compareTo(BigDecimal.ZERO) == 0) {
            return k == 0 ? BigDecimal.ONE : BigDecimal.ZERO;
        }

        if (p.compareTo(BigDecimal.ONE) == 0) {
            return k == n ? BigDecimal.ONE : BigDecimal.ZERO;
        }

        BigDecimal binomialLn = SpecialFunctions.GammaLn(new BigDecimal(n + 1))
                .subtract(SpecialFunctions.GammaLn(new BigDecimal(k + 1)))
                .subtract(SpecialFunctions.GammaLn(new BigDecimal(n - k + 1)));

        return MathEx.Exp(binomialLn
                .add(new BigDecimal(k).multiply(MathEx.Log(p)))
                .add(new BigDecimal(n - k).multiply(MathEx.Log(BigDecimal.ONE.subtract(p)))));
    }

    public static BigDecimal CDF(BigDecimal p, int n, BigDecimal x) {
        if (x.compareTo(BigDecimal.ZERO) < 0) {
            return BigDecimal.ZERO;
        }

        if (x.compareTo(new BigDecimal(n)) > 0) {
            return BigDecimal.ONE;
        }

        BigDecimal k = new BigDecimal(Math.floor(x.doubleValue()));
        return SpecialFunctions.BetaRegularized(new BigDecimal(n - k.intValue()), new BigDecimal(k.intValue() + 1), BigDecimal.ONE.subtract(p));
    }
}

File: csharp/ToolGood.Algorithm/MathNet/Distributions/Binomial.cs
using System;

namespace ToolGood.Algorithm.MathNet.Numerics.Distributions
{
    sealed class Binomial
    {
        public static decimal PMF(decimal p, int n, int k)
        {
            if (k < 0 || k > n) {
                return 0.0m;
            }

            if (p == 0.0m) {
                return k == 0 ? 1.0m : 0.0m;
            }

            if (p == 1.0m) {
                return k == n ? 1.0m : 0.0m;
            }

            return MathEx.Exp(SpecialFunctions.BinomialLn(n, k) + (k * MathEx.Log(p)) + ((n - k) * MathEx.Log(1.0m - p)));
        }

        public static decimal CDF(decimal p, int n, decimal x)
        {
            if (x < 0.0m) {
                return 0.0m;
            }

            if (x > n) {
                return 1.0m;
            }

            if (p <= 0.0m) {
                // p=0 时全部概率集中在 k=0
                return 1.0m;
            }

            if (p >= 1.0m) {
                // p=1 时全部概率集中在 k=n
                return x >= n ? 1.0m : 0.0m;
            }

            if (x >= n) {
                // x=n 时 n-k=0, BetaRegularized 内部会出现 0/0
                return 1.0m;
            }

            decimal k = Math.Floor(x);
            return SpecialFunctions.BetaRegularized(n - k, k + 1, 1 - p);
        }
    }
}
